library/x509_crt.c

```diff
@@ -1125,7 +1139,10 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
     p = filename + len;
     filename[len++] = '*';
 
-    w_ret = MultiByteToWideChar( CP_ACP, 0, filename, (int)len, szDir,
+    if ( FAILED ( SizeTToInt( len, &lengthAsInt ) ) )
+        return( MBEDTLS_ERR_X509_FILE_IO_ERROR );
+
+    w_ret = MultiByteToWideChar( CP_ACP, 0, filename, lengthAsInt, szDir,
                                  MAX_PATH - 3 );
```

Comment:
Preexisting: at some point someone is going to complain that `CP_ACP` is obsolescent. We should add a warning to state that this code depends on using a unibyte code page, because the output buffer size is in characters, not bytes. Changing to e.g. `CP_UTF8` would require more complex code here. See https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx

Reply:
Given the work @irwir has been doing on PR #1427, I think I'd rather this was raised as a bug, rather than a comment in the code.

If you agree, could you please raise the bug?

Reply:
I don't see any functional bug here, just a maintenance issue that could lead to a bug in the future. This needs to be a comment in the code. What would be the bug? To switch to `CP_UTF8`? I don't know if it's the right thing. For all I know, `CP_ACP` is fine because we want to treat the file name as just some string to be passed (semantically) unchanged and we don't care about encoding.

Reply:
It seems to me that #1427 would bypass this issue. I'd still rather have the comment, even in the backports.

Reply:
I think I should have said raise it as an issue, not a bug, and that implies a fault.

I'm not sure my point was obvious enough to I'll explain it again. @irwir has created a PR to change the strings that touch the platform from `char`'s to `TCHAR`'s to allow portability between ASCII/UTF-8 and Unicode, including changing interface code so the public interface on Windows can accept `TCHAR` strings.

I do think we need to consider how we interface with Windows and what we should do about this, and consider if and how we address the issue.

My suggestion of raising an issue was really about that.

Anyway - I'll take your point and add a comment.

Reply:
I didn't want to change this code more than I had to when I originally added this, but since you've brought it up, it looks like the whole point of these conversions is to call FindFirstFileW and FindNextFileW, and then to convert their results back to narrow character strings to be used.

Another possibility is to dispense with wide character strings entirely, since the mbedTLS API always takes narrow character strings, and use the narrow character string Windows equivalents: the FindFirstFileA and FindNextFileA functions and the WIN32_FIND_DATAA type. Then the platform will do all the string conversions internally and this code can only deal in narrow character strings.

If you decide to move to a TCHAR API approach then all of this is moot, of course, but for however long you stick with narrow character strings, you might consider this.

Reply:
> this code depends on using a unibyte code page

Is 'unibyte' word about one byte being one character?
In that case, ANSI is no more unibyte than UTF-8. Just remember ANSI code pages for Oriental languages.
Whereever needed, Windows provide two APIs for the same function: one ANSI, and one Unicode.
No conversion should be needed when strings and API were both of ANSI kind.

Reply:
Yes, “unibyte” means that 1 character = 1 byte. Thanks @irwir, I didn't know that ANSI could be multibyte (I'm not familiar with Windows APIs at all). Given that `szDir` is a `WCHAR` array, is the size correct? In other words, does an ANSI character always fit in a `WCHAR`?

Reply:
> does an ANSI character always fit in a WCHAR?

That could be expected in most cases, but not guaranteed because `WCHAR` implies Unicode with its peculiarities.
`MAX_PATH` dedermines the number of array elements, not the number of characters (see description of  `WIN32_FIND_DATA` structure). And, as said, conversion to wide character is not needed.

Reply:
> Another possibility is to dispense with wide character strings entirely, since the mbedTLS API always takes narrow character strings

The cryptographic part actually uses bytes and does not care about characters.
Characters appear when API refers to path/file names. As I was told, it was supposed to be  in UTF-8 encoding, and that is not what 'narrow' strings were supposed to mean in Windows code.